### PR TITLE
Add ability to override centrally defined package version

### LINKED ISSUE #11516: [Feature]: Allow overriding a centrally defined package version
### NuGet Product(s) Involved

NuGet.exe, MSBuild.exe, dotnet.exe

### The Elevator Pitch

Central Package Management (CPM) allows users to specify versions for their packages in one place.  However, there are edge cases where a project might want to use a different version.  For example, a prototype project could need a newer version than a repo is using so it would opt into newer versions.

There is an existing implementation, Microsoft.Build.CentralPackageVersions, which uses a `VersionOverride` attribute to indicate that a `<PackageReference />` should use a different version than is centrally defined.

This design has a few advantages:
1. Makes it obvious when a project is using its own package version(s)
2. Ideally, an entire repo would use the same package version which is why users would define it in one place.  Users who want to use a different version would need to fully understand the ramifications of specifying the `VersionOverride` attribute.
3. Migration from Microsoft.Build.CentralPackageVersions would be easier.

Other options for users that would still be possible are:
1. Specify a `<PackageVersion />` in a project
2. Define an MSBuild property that represents the package version and set it in a project

Since MSBuild essentially lets users do anything they want, the built-in mechanism for overriding a package version should be simple and well documented.  

# Design

A user would specify package versions in `Directory.Packages.props`:
```xml
<Project>
  <ItemGroup>
    <PackageVersion Include="PackageA" Version="1.0.0" />
    <PackageVersion Include="PackageB" Version="2.0.0" />
  </ItemGroup>
<Project>
```

A project would override the version with the `VersionOverride` metadata on a `<PackageReference />` item:

```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>net5.0</TargetFramework>
  </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="PackageA" VersionOverride="3.0.0" />
  </ItemGroup>
<Project>
```

In this case, the project's restore graph would resolve `PackageA` to version `3.0.0`.  Any project that references it would also get that version and a user would be responsible for handling an unresolved conflicts.

Finally, a repo owner should be able to disable the ability for developers to override package version.  This would be used for instance if someone wanted to ensure that all package versions are unified.  This would be possible by setting the MSBuild property `EnablePackageVersionOverride` to `false` in a project or import like Directory.Build.props:

```xml
<Project>
  <PropertyGroup>
    <EnablePackageVersionOverride>false</EnablePackageVersionOverride>
  </PropertyGroup>
<Project>
```

When this is disabled, specifying a `VersionOverride` on a `<PackageReference />` would result in a restore error indicating that he feature is disabled.

### Additional Context and Details

_No response_

### PR DESCRIPTION
**Pull request by jeffkl**
## Bug

Fixes: https://github.com/NuGet/Home/issues/11516

Regression? Last working version:

## Description
Add the ability for `<PackageReference />` items to specify a `VersionOverride` attribute to override the centrally defined package version.  Also adds an MSBuild property to disable the functionality if a repo owner does not want version overrides to be introduced.

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [x] Automated tests added
  - **OR**
  
  - [ ] Test exception
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [x] Documentation PR or issue filled https://github.com/NuGet/docs.microsoft.com-nuget/issues/2674
  - **OR**
  - [ ] N/A

### CODE REVIEW

**Comment by nkolev92** [comment]
I'll review this, but I really think we should have a review process with the community/customer for this.

**Comment by jeffkl** [comment]
> I'll review this, but I really think we should have a review process with the community/customer for this.

@JonDouglas has hosted a few customer discussions and we feel confident that this implementation covers what was asked for.  I'm also hoping that any customer who is interested will comment on this PR or the associated issue about the implementation.

**Comment by nkolev92** [comment]
> @JonDouglas has hosted a few customer discussions and we feel confident that this implementation covers what was asked for. I'm also hoping that any customer who is interested will comment on this PR or the associated issue about the implementation.

Fair, but we should make sure the documentation/specs in the Home repo are updated.

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs:872]: 👍

File: test/NuGet.Clients.Tests/NuGet.SolutionRestoreManager.Test/VsSolutionRestoreServiceTests.cs
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.SolutionRestoreManager.Test/VsSolutionRestoreServiceTests.cs]: use explicit type

File: test/NuGet.Clients.Tests/NuGet.SolutionRestoreManager.Test/VsSolutionRestoreServiceTests.cs
Comment (donnie-msft) [inline test/NuGet.Clients.Tests/NuGet.SolutionRestoreManager.Test/VsSolutionRestoreServiceTests.cs]: Could do `Assert.Equals` for a one-liner

File: test/NuGet.Core.Tests/NuGet.ProjectModel.Test/DependencyGraphSpecTests.cs
Comment (donnie-msft) [inline test/NuGet.Core.Tests/NuGet.ProjectModel.Test/DependencyGraphSpecTests.cs]: nit: all these `versionOverride`'s are indented differently

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs]:
```suggestion
        private static VersionRange GetVersionOverride(PackageReference reference)
```

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (donnie-msft) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: Can this not be a double negative?

File: src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs
Comment (donnie-msft) [inline src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs]:
```suggestion
        /// Determines the current settings for central package management for the specified project.
```

**Comment by jeffkl** [comment]
This is now ready for review based on the proposal at https://github.com/NuGet/Home/pull/11569

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs
Comment (erdembayar) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/ProjectServices/VsManagedLanguagesProjectSystemServices.cs:254]: Other methods in same cs file have following, maybe we should follow it.
```Suggestion
      {
         Assumes.Present(reference);
```

File: src/NuGet.Core/NuGet.Build.Tasks.Console/ExtensionMethods.cs
Comment (erdembayar) [inline src/NuGet.Core/NuGet.Build.Tasks.Console/ExtensionMethods.cs:21]: Do we really need IsPropertyFalse if we already have `IsPropertyTrue` method? Why don't we just use `!IsPropertyTrue` instead?

File: test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs
Comment (erdembayar) [inline test/NuGet.Clients.Tests/NuGet.PackageManagement.VisualStudio.Test/ProjectSystems/LegacyPackageReferenceProjectTests.cs:876]: How about `FALSE`?

File: src/NuGet.Core/NuGet.Build.Tasks.Console/ExtensionMethods.cs
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Build.Tasks.Console/ExtensionMethods.cs:21]: Great question.  As a best practice, you shouldn't introduce MSBuild properties if you can rely on a default.

For example:

```xml
<PropertyGroup>
  <IsSomethingEnabled>true</IsSomethingEnabled>
</PropertyGroup>
```

As long as you document this property, there's no reason to allocate a string to track the fact that its set to `true`.  Instead, you only have to check if the user turned it off by setting it to `false`.

We've added other properties with their values explicitly set to `true` and then check if the value is still `true`.  Going forward, I'd like us to adopt this pattern of not introducing a property and instead checking if the user has set it to `false`.  You also can't negate `IsPropertyTrue()` since we're just comparing strings and the only supported value is `false`, anything else is considered `true`.

Comment (erdembayar) [reply src/NuGet.Core/NuGet.Build.Tasks.Console/ExtensionMethods.cs:21]: Thank you for context.

File: src/NuGet.Core/NuGet.LibraryModel/LibraryDependency.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.LibraryModel/LibraryDependency.cs:158]: Do we ever use the overriden information? 

Would we be better off just adding a property that tells us we have an override instead?

Comment (nkolev92) [reply src/NuGet.Core/NuGet.LibraryModel/LibraryDependency.cs:158]: It basically seems less error to do it from the onset instead of replacing it here.

**Review by nkolev92 — COMMENTED** [review]
First pass, looks reasonable to me. 

One small question, but I think the general idea is the right one.

It seems like you found all the right places :D

File: src/NuGet.Core/NuGet.LibraryModel/LibraryDependency.cs
Comment (jeffkl) [reply src/NuGet.Core/NuGet.LibraryModel/LibraryDependency.cs:158]: We need to be able to distinguish between when a user specified a `Version` (to emit an error and just `VersionOverride`.  At the moment, that information is gathered from a source (project system, MSBuild) and we create a `ProjectSpec`.  When processing the `ProjectSpec` as part of restore is when this information is processed.

To do it the other way around, I'd need to update the code in the following data gathering code paths which construct the `ProjectSpec`:
* Legacy project system
* CPS
* MSBuild/dotnet GenerateRestoreGraph
* MSBuild/dotnet Static graph

It felt simpler at the time to leave those systems as just gathering the data for processing later.  If you feel strongly, I can update the code to apply the versions when gathering the data and emit any errors/warnings there.

Comment (nkolev92) [reply src/NuGet.Core/NuGet.LibraryModel/LibraryDependency.cs:158]: Oh I understand why we need the information that the version has been overriden. 

My suggestion is instead of Version and VersionOverride to have, Version, where it's the effective version and OriginalVersion, or a bool to indicate the the version has been changed, such as VersionOverriden. 

That way you don't need to rewrite the versionn range and you're less likely to get into funny business because the majority of the restore code still just uses, Version.

**Comment by nkolev92** [comment]
@jeffkl Ready for rereview right?

You might want to rebase.

**Comment by jeffkl** [comment]
> You might want to rebase.

Done!

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets]: Usually the configuration properties are named starting with `Restore`. 
Do you think that should be the case here as well?

File: src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs
Comment (nkolev92) [inline src/NuGet.Core/NuGet.Commands/RestoreCommand/RestoreCommand.cs]: Errors are from 000-499, warnings are from 500-999.

File: src/NuGet.Core/NuGet.LibraryModel/LibraryDependency.cs
Comment (nkolev92) [reply src/NuGet.Core/NuGet.LibraryModel/LibraryDependency.cs:158]: Resolving this.

File: test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs]: The CreateFolderFeedV3Async method takes params.

File: test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs
Comment (nkolev92) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs:2097]: You might have an easier time with these if you get them through the GetPackageSpec. 

You'd reduce the the potential inconsistencies in the package spec.s

**Review by nkolev92 — COMMENTED** [review]
The warning code needs to tweaked. 
Looks great beyond that.

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets]: This is a great question, I wasn't sure about the name of this property either.  I did notice that some of the properties are prefixed with `Restore` and I think that makes sense if they directly tweak settings for restore.  In this case, its more of a "feature of the restore graph" or whatever and not really a knob that controls restore itself.

For example, the property to disable/enable central package management is `ManagePackageVersionsCentrally`.  The property to tell restore if interactivity is enabled/disabled is `NuGetInteractive`.  So I wasn't sure what to name this thing.

 I don't feel strongly one way or the other.  Here are some potential names:
* `EnablePackageVersionOverride`
* `AllowPackageVersionOverride`
* `RestoreEnablePackageVersionOverride`
* `RestoreAllowPackageVersionOverride`

What does everyone think?

File: test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs
Comment (jeffkl) [reply test/NuGet.Core.Tests/NuGet.Commands.Test/RestoreCommandTests.cs:2097]: I copy/pasted this pattern from existing tests.  I'm going to leave it as-is for now, maybe I can clean up these tests later

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets]: Honestly, I'm comfortable with either of the 4 :D 

I like the `RestoreEnablePackageVersionOverride`, or `EnablePackageVersionOverride`, even if `Allow` is technically more correct.

@zivkan thoughts?

**Review by nkolev92 — APPROVED** [review]
LGTM. 

Couple of things before merging: 

- Property name for enabling the feature.
- Fill out the template. Likely that means creating docs issues for the log codes and the overall override feature.

**Comment by jeffkl** [comment]
> Fill out the template. Likely that means creating docs issues for the log codes and the overall override feature.

Thanks for the reminder:  https://github.com/NuGet/docs.microsoft.com-nuget/issues/2674

**Review by erdembayar — APPROVED** [review]

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (zivkan) [inline src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: Can you split this across multiple lines.  When debugging it can be very useful to inspect the output of `GetPropertyValueAsync`, or to step into `IsFalse` without stepping into `GetPropertyValueAsync`.

Personally, I think that calling a method from inside another method's parameter list should be against our coding guidelines. Variables are free.

File: src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs
Comment (zivkan) [inline src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:280]: Out of scope, but something to think about, Nuget has really terrible error handling, often crashing/failing restore at first error, rather than trying to limp along to a gatepoint to collect all errors. In this case, if the value for `VersionOverride` can't be parsed as a version range, this project will fail restore immediately, and if there are any other syntax errors, they won't get reported to the customer. They have to fix the errors one by one, retrying restore, getting another single error, then repeat.

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
Comment (zivkan) [reply src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets]: I don't know. Since MSBuild properties could be for anything, NuGet, or the compiler, or output directories, or other things, I see value in a naming convention. On the other hand, many features like implicit usings, just have a minimal name, assuming the customer understands the context for which it's being used, so `ManagedPackageVersionsCentrally` and `CentralPackageVersionOverrideEnabled` seems to fit.

Perhaps if you want another opinion, try asking people on the .NET SDK team.

File: src/NuGet.Core/NuGet.Common/PublicAPI/net45/PublicAPI.Unshipped.txt
Comment (zivkan) [inline src/NuGet.Core/NuGet.Common/PublicAPI/net45/PublicAPI.Unshipped.txt]: I know you're just adding a related method to an existing public class, so I'm just venting about NuGet's public APIs.  I don't feel like we should be a utility library for developers using the NuGet SDK to parse MSBuild properties. If there's value in that, I'd expect MSBuild themselves to provide such helper methods.

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
Comment (nkolev92) [reply src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets]: I actually like CentralPackageVersionOverrideEnabled, because it's more explicit, but the current proposal is EnablePackageVersionOverride.

**Review by khaledxGo — APPROVED** [review]
Hello world

File: src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Build.Tasks/NuGet.targets]: Okay I'm going to change the name to `CentralPackageVersionOverrideEnabled`

File: src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs
Comment (jeffkl) [reply src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Projects/LegacyPackageReferenceProject.cs]: Yeah sorry, was copy/pasted so I'm going to fix the method above it as well!

File: src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:280]: I saw this too.  At first I was hoping something else happened like the APIs would catch it and present a better error message.  So I tried it out and saw the behavior today if a version is invalid, our code just gives up immediately.  Definitely not the best user experience.  Do we have an existing issue open to fix that?

File: src/NuGet.Core/NuGet.Common/PublicAPI/net45/PublicAPI.Unshipped.txt
Comment (jeffkl) [reply src/NuGet.Core/NuGet.Common/PublicAPI/net45/PublicAPI.Unshipped.txt]: I agree, I didn't like this either. We don't make this class available through `InternalsVisibleTo` either.  But we do have shared code that's compiled in so I've moved this to `EqualityUtility` instead.  What do you think?

File: src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs
Comment (zivkan) [reply src/NuGet.Core/NuGet.Build.Tasks.Console/MSBuildStaticGraphRestore.cs:280]: Not that I'm aware of, no. I think it would be great for customers, but I fear how much effort it would be to implement.

**Review by zivkan — APPROVED** [review]

File: src/NuGet.Core/NuGet.LibraryModel/LibraryDependency.cs
Comment (kartheekp-ms) [reply src/NuGet.Core/NuGet.LibraryModel/LibraryDependency.cs:158]: > My suggestion is instead of Version and VersionOverride to have, Version, where it's the effective version and OriginalVersion, or a bool to indicate the the version has been changed, such as VersionOverriden.

I am thinking the same. Given the comment has been resolved, I am curious to learn more incase if I missed any context.

**Comment by kartheekp-ms** [comment]
one question regarding tests - Do we have any tests in this PR that test to make sure [`GeneratePathProperty`](https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#generatepathproperty) functionality is working as expected when `VersionOverride` is specified? Sorry if a test is already added in this PR and I missed it.

File: test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs
Comment (marcin-krystianc) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs]:
```suggestion
        /// Verifies that <see cref="MSBuildRestoreUtility.GetDependencySpec(IEnumerable{IMSBuildItem})" /> applies version overrides correctly depending on whether or not central package management is enabled.
```

File: test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs
Comment (marcin-krystianc) [inline test/NuGet.Core.Tests/NuGet.Commands.Test/MSBuildRestoreUtilityTests.cs]:
```suggestion
        /// <param name="isCentralPackageManagementEnabled"><c>true</c> if central package management is enabled, otherwise <c>false</c>.</param>
```

**Comment by jeffkl** [comment]
> one question regarding tests - Do we have any tests in this PR that test to make sure [GeneratePathProperty](https://docs.microsoft.com/en-us/nuget/consume-packages/package-references-in-project-files#generatepathproperty) functionality is working as expected when VersionOverride is specified?

This shouldn't matter as the version is overridden when the graph is resolved and GeneratePathProperty is used much later when writing out the MSBuild files which uses whatever version was in the final graph.  The tests are making sure that the correct version is resolved so I think that should be covering the scenario.

**Pull request merged**
